Fix kubeconfig-sync sometimes producing multiple identical entities

Nokel81 opened the pull request:
- This happens when there is an overlapping sync of a folder and a file
  within that folder both producing the same entities

- This also fixes the LED not showing up on some KubernetesCluster
  instances because of this duplication the connected status was being
  set on the first instance, but renderer would use the second (because
  JS's Map constructor deduplicates its constructor initializer by
  taking the last)

- Added some tests to cover this case

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix kubeconfig-sync sometimes producing multiple identical entities

- This happens when there is an overlapping sync of a folder and a file
  within that folder both producing the same entities

- This also fixes the LED not showing up on some KubernetesCluster
  instances because of this duplication the connected status was being
  set on the first instance, but renderer would use the second (because
  JS's Map constructor deduplicates its constructor initializer by
  taking the last)

- Added some tests to cover this case

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Move initializing ClusterManager into a runable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

aleksfront reviewed (commented):
I can't reproduce the issue mentioned in the description:

> This happens when there is an overlapping sync of a folder and a file
within that folder both producing the same entities

Tested with 2 different file/folder structures:
```
configfolder -> config
             -> deepfolder -> config
``` 
and
```
configfolder -> config
             -> config 2
```

all of the config files contained same minikube cluster (copy pasted kubeconfig from one file to another)

In my opinion Lens seems work as expected treating these clusters as separate ones. Both of clusters switched okay with proper LED lights without any sync problem. Example:


https://user-images.githubusercontent.com/9607060/187372472-2bb205e1-306b-4060-be97-4bfcf21d7e92.mov

Iku-turso commented on src/common/logger/child-logger.injectable.ts:
```diff
@@ -0,0 +1,37 @@
+  instantiate: (di, prefix): Logger => {
+    const logger = di.inject(loggerInjectable);
+
+    return {
```
Note: it's a questionable practise here and elsewhere to bundle all these responsibilities together.

Iku-turso replied:
Discussion handled IRL.

Iku-turso commented on src/common/logger/child-logger.injectable.ts:
```diff
@@ -0,0 +1,37 @@
+import loggerInjectable from "../logger.injectable";
+
+const childLoggerInjectable = getInjectable({
+  id: "child-logger",
```
Proposed name: "prefixed-logger" :)

Iku-turso commented on src/main/catalog-sources/__test__/kubeconfig-sync.test.ts:
```diff
@@ -151,7 +164,7 @@ describe("kubeconfig-sync.source tests", () => {
 
       fs.writeFileSync(filePath, contents);
```
Opportunity to get rid of another `mock-fs` ;) Fix if feeling boy-scoutish :)

While we are at it, maybe these technical unit tests could be more awesome as part of behavioural tests instead. WDYT? :)

Iku-turso replied:
Discussion handled IRL.

Iku-turso commented on src/common/fs/watch.injectable.ts:
```diff
@@ -0,0 +1,14 @@
+import { getInjectable } from "@ogre-tools/injectable";
+import { watch } from "chokidar";
+
+const watchInjectable = getInjectable({
```
Nit: this foreign 3rd party abstraction is accessed directly from our domestic code, hard-coupling us to chokidar exactly. We could just as well invent our own abstraction that would be simple, stable and controlled in single place.

This may not seem super meaningful here, but as a policy, not letting foreign abstractions bleed in domestic code can sure be a life-saver, and a good investment on top of that.

Iku-turso replied:
Discussion handled IRL.

Iku-turso commented on src/main/start-main-application/runnables/kube-config-sync/add-source.injectable.ts:
```diff
@@ -0,0 +1,25 @@
+/**
```
As modern code like this emerges, it would be great if it did so accompanied by those darn behavioural tests ;) WDYT?

Iku-turso replied:
Discussion handled IRL.

Iku-turso reviewed (changes requested):
Discuss: Add or migrate to behavioural tests? Are you too starting to have the feeling that PR's that don't pull their weight in contributions to unit testing don't kinda feel up-to-par anymore? No? :)

Nokel81 committed:
Change childLogger to prefixedLogger

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Remove duplicate watchInjectable

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix wrong import in test

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Switch to overriding dependencies to fix test flakiness

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Make Runnable and RunnableSync require ids

- Allows for better error messages related to mismatched runAfter's so
  that mismatched injection tokens are easy to fix

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Fix crash related to mimatched runnable injection tokens

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Improve error message for mismatched runnables

Signed-off-by: Sebastian Malton <sebastian@malton.name>

Nokel81 committed:
Add IDs to runnables

Signed-off-by: Sebastian Malton <sebastian@malton.name>
